File: telegram-bot-sdk/haskell/src/Agent.hs
{-# LANGUAGE ExistentialQuantification #-}
-- | Agent framework for the Devflow bot.
-- Provides a Command typeclass with dynamic dispatch, a context type,
-- and an Ollama-powered fallback for unrecognised commands.
module Agent
  ( -- * Command registry
    CommandRegistry (..)
  , emptyRegistry
  , registerCommand
  , lookupCommand
  , registeredNames
    -- * Dispatch
  , dispatch
  , dispatchWithFallback
    -- * Agent context
  , AgentContext (..)
  , mkAgentContext
    -- * Fallback
  , ollamaFallback
  , systemPrompt
  ) where

import           Control.Exception       (SomeException, try)
import           Data.Map.Strict         (Map)
import qualified Data.Map.Strict         as Map
import           Data.Text               (Text)
import qualified Data.Text               as T
import           Lib.Ollama              (OllamaConfig, OllamaMessage (..),
                                          OllamaRole (..), generate)
import           Types                   (BotContext (..), BotError (..),
                                          BotResponse (..), ChatId,
                                          Command (..), CommandName (..),
                                          OllamaError (..), ResponseKind (..))

-- ---------------------------------------------------------------------------
-- CommandRegistry
-- ---------------------------------------------------------------------------

-- | A heterogeneous registry of commands keyed by name.
-- We existentially wrap each command so they can be stored together.
data AnyCommand = forall cmd. Command cmd => AnyCommand cmd

newtype CommandRegistry = CommandRegistry
  { crCommands :: Map CommandName AnyCommand
  }

emptyRegistry :: CommandRegistry
emptyRegistry = CommandRegistry Map.empty

registerCommand :: Command cmd => cmd -> CommandRegistry -> CommandRegistry
registerCommand cmd (CommandRegistry m) =
  CommandRegistry (Map.insert (commandName cmd) (AnyCommand cmd) m)

lookupCommand :: CommandName -> CommandRegistry -> Maybe AnyCommand
lookupCommand name (CommandRegistry m) = Map.lookup name m

registeredNames :: CommandRegistry -> [CommandName]
registeredNames (CommandRegistry m) = Map.keys m

-- ---------------------------------------------------------------------------
-- AgentContext
-- ---------------------------------------------------------------------------

-- | Enhanced context passed to the dispatch pipeline.
data AgentContext = AgentContext
  { acBotCtx    :: BotContext
  , acRegistry  :: CommandRegistry
  , acOllama    :: OllamaConfig
  , acUseFallback :: Bool   -- ^ allow Ollama fallback for unknown commands
  }

mkAgentContext
  :: BotContext
  -> CommandRegistry
  -> OllamaConfig
  -> AgentContext
mkAgentContext ctx reg oCfg = AgentContext
  { acBotCtx      = ctx
  , acRegistry    = reg
  , acOllama      = oCfg
  , acUseFallback = True
  }

-- ---------------------------------------------------------------------------
-- Dispatch
-- ---------------------------------------------------------------------------

-- | Execute a named command.  Returns Left (CommandNotFound name) if the
-- command is not in the registry.
dispatch
  :: AgentContext
  -> CommandName
  -> IO (Either BotError [BotResponse])
dispatch ac name =
  case lookupCommand name (acRegistry ac) of
    Nothing -> pure (Left (CommandNotFound name))
    Just (AnyCommand cmd) -> commandExecute cmd (acBotCtx ac)

-- | Execute a named command; if unknown and fallback is enabled,
-- forward the raw message text to Ollama.
dispatchWithFallback
  :: AgentContext
  -> CommandName
  -> IO (Either BotError [BotResponse])
dispatchWithFallback ac name =
  case lookupCommand name (acRegistry ac) of
    Just (AnyCommand cmd) -> commandExecute cmd (acBotCtx ac)
    Nothing
      | acUseFallback ac ->
          ollamaFallback (acOllama ac) (acBotCtx ac) (ctxRawText (acBotCtx ac))
      | otherwise ->
          pure (Left (CommandNotFound name))

-- ---------------------------------------------------------------------------
-- Ollama fallback
-- ---------------------------------------------------------------------------

-- | The system prompt used when forwarding messages to Ollama.
systemPrompt :: Text
systemPrompt = T.unlines
  [ "You are Devflow, an AI assistant embedded in a Telegram bot."
  , "You help development teams manage sprints, issues, and backlogs."
  , "Be concise, helpful, and direct. Prefer bullet lists when appropriate."
  , "When asked about available commands, mention /sprint, /item, /board,"
  , "/issue, /bug, /resolve, /issues, and /ask."
  , "Do not make up information. If you don't know, say so."
  ]

-- | Send the raw user text to Ollama and wrap the response.
ollamaFallback
  :: OllamaConfig
  -> BotContext
  -> Text              -- ^ raw message text
  -> IO (Either BotError [BotResponse])
ollamaFallback oCfg ctx rawText = do
  result <- try (generate oCfg rawText (Just systemPrompt))
              :: IO (Either SomeException (Either OllamaError Text))
  case result of
    Left ex -> pure . Left . InternalError . T.pack $ show ex
    Right (Left ollamaErr) -> handleOllamaError ctx ollamaErr
    Right (Right responseText) ->
      pure . Right $
        [ BotResponse
            { respChatId  = ctxChatId ctx
            , respText    = responseText
            , respKind    = MarkdownResponse
            , respReplyTo = Nothing
            }
        ]

handleOllamaError :: BotContext -> OllamaError -> IO (Either BotError [BotResponse])
handleOllamaError ctx ollamaErr =
  case ollamaErr of
    OllamaConnectionError msg ->
      -- Graceful degradation: tell the user Ollama is unavailable
      pure . Right $
        [ BotResponse
            { respChatId  = ctxChatId ctx
            , respText    = "AI assistant is currently unavailable. " <>
                            "Please use /help to see available commands."
            , respKind    = TextResponse
            , respReplyTo = Nothing
            }
        ]
    OllamaTimeoutError ->
      pure . Right $
        [ BotResponse
            { respChatId  = ctxChatId ctx
            , respText    = "AI assistant timed out. Please try again."
            , respKind    = TextResponse
            , respReplyTo = Nothing
            }
        ]
    other ->
      pure (Left (OllamaErr other))

-- ---------------------------------------------------------------------------
-- Dynamic dispatch table builder
-- ---------------------------------------------------------------------------

-- | Build a help text listing all registered commands.
registryHelpText :: CommandRegistry -> Text
registryHelpText reg =
  let names = registeredNames reg
      formatName (CommandName n) = "  /" <> n
      lines' = map formatName names
  in "*Available Commands:*\n" <> T.unlines lines'

-- | Return a registry help response.
helpResponse :: AgentContext -> [BotResponse]
helpResponse ac =
  [ BotResponse
      { respChatId  = ctxChatId (acBotCtx ac)
      , respText    = registryHelpText (acRegistry ac)
      , respKind    = MarkdownResponse
      , respReplyTo = Nothing
      }
  ]

File: telegram-bot-sdk/rust/src/agent.rs
//! Agent framework — command routing with Ollama LLM fallback.
//!
//! The `Agent` struct is the central dispatcher.  It:
//! 1. Tries to match the input against the `CommandRegistry`.
//! 2. If no command matches and Ollama is configured, calls the LLM.
//! 3. Otherwise returns a "not understood" response.
//!
//! `AgentCommand` is the richer command trait used here (carries `Arc<AppState>`
//! explicitly instead of storing it on the struct).

use std::collections::HashMap;
use std::sync::Arc;
use async_trait::async_trait;
use tracing::{debug, info, warn};

use crate::error::{BotError, Result};
use crate::integrations::ollama::{OllamaClient, OllamaConfig};
use crate::state::AppState;
use crate::types::{Context, Response};

// ---------------------------------------------------------------------------
// AgentCommand trait
// ---------------------------------------------------------------------------

/// Like `Command` but receives `Arc<AppState>` at dispatch time rather than
/// storing it.  Use this for commands that need to be registered in the agent
/// registry without tying them to a specific `AppState` at construction time.
#[async_trait]
pub trait AgentCommand: Send + Sync {
    fn name(&self) -> &str;
    fn description(&self) -> &str;
    fn aliases(&self) -> &[&str] { &[] }

    /// Returns `true` if this command can handle the input token.
    /// Default: exact match on `name()` or any alias.
    fn can_handle(&self, token: &str) -> bool {
        let token = token.trim_start_matches('/');
        self.name() == token || self.aliases().contains(&token)
    }

    async fn execute(
        &self,
        ctx: &Context,
        args: &[&str],
        state: Arc<AppState>,
    ) -> Result<Response>;
}

// ---------------------------------------------------------------------------
// CommandRegistry
// ---------------------------------------------------------------------------

/// Holds `AgentCommand` implementations indexed by all their names and aliases.
#[derive(Default)]
pub struct CommandRegistry {
    commands: HashMap<String, Arc<dyn AgentCommand>>,
}

impl CommandRegistry {
    pub fn new() -> Self {
        Self::default()
    }

    /// Register a command by its name and all aliases.
    pub fn register(&mut self, cmd: Arc<dyn AgentCommand>) {
        let name = cmd.name().to_owned();
        let aliases: Vec<String> = cmd.aliases().iter().map(|&s| s.to_owned()).collect();
        self.commands.insert(name, Arc::clone(&cmd));
        for alias in aliases {
            self.commands.insert(alias, Arc::clone(&cmd));
        }
    }

    /// Look up a command by the raw first token (leading `/` stripped).
    pub fn get(&self, token: &str) -> Option<Arc<dyn AgentCommand>> {
        let key = token.trim_start_matches('/');
        self.commands.get(key).cloned()
    }

    pub fn len(&self) -> usize {
        self.commands.len()
    }

    pub fn is_empty(&self) -> bool {
        self.commands.is_empty()
    }

    /// Returns sorted unique command names (no aliases).
    pub fn names(&self) -> Vec<&str> {
        let mut names: Vec<&str> = self.commands.keys().map(String::as_str).collect();
        names.sort_unstable();
        names
    }
}

// ---------------------------------------------------------------------------
// AgentConfig
// ---------------------------------------------------------------------------

/// Tuning knobs for the `Agent`.
#[derive(Debug, Clone)]
pub struct AgentConfig {
    /// Prepended to every Ollama prompt.
    pub system_prompt: String,
    /// If `true`, unrecognised messages are silently ignored (no "I don't understand").
    pub silent_unknown: bool,
    /// Max context window fed to Ollama (number of recent session entries).
    pub max_context_entries: usize,
}

impl Default for AgentConfig {
    fn default() -> Self {
        Self {
            system_prompt: "You are DevFlow, an AI assistant specialising in agile project \
                            management, sprint planning, and issue tracking. Keep answers concise \
                            and actionable."
                .to_owned(),
            silent_unknown: false,
            max_context_entries: 6,
        }
    }
}

// ---------------------------------------------------------------------------
// Agent
// ---------------------------------------------------------------------------

/// Top-level dispatcher.
///
/// Constructed once and stored as `Arc<Agent>` in the bot's dependency
/// injection container.
pub struct Agent {
    registry: CommandRegistry,
    ollama: Option<OllamaClient>,
    state: Arc<AppState>,
    config: AgentConfig,
}

impl Agent {
    // -----------------------------------------------------------------------
    // Construction
    // -----------------------------------------------------------------------

    /// Build an `Agent` with no commands and no Ollama client.
    pub fn new(state: Arc<AppState>) -> Self {
        Self {
            registry: CommandRegistry::new(),
            ollama: None,
            state,
            config: AgentConfig::default(),
        }
    }

    /// Builder: enable Ollama with a custom config.
    pub fn with_ollama(mut self, config: OllamaConfig) -> Result<Self> {
        self.ollama = Some(OllamaClient::new(config)?);
        Ok(self)
    }

    /// Builder: enable Ollama reading from environment variables.
    pub fn with_ollama_from_env(mut self) -> Self {
        match OllamaClient::from_env() {
            Ok(client) => {
                info!("Ollama LLM enabled at {}", client.base_url());
                self.ollama = Some(client);
            }
            Err(e) => {
                warn!("Ollama unavailable, LLM fallback disabled: {}", e);
            }
        }
        self
    }

    /// Builder: apply a custom agent config.
    pub fn with_config(mut self, config: AgentConfig) -> Self {
        self.config = config;
        self
    }

    /// Builder: register an `AgentCommand`.
    pub fn register(mut self, cmd: Arc<dyn AgentCommand>) -> Self {
        self.registry.register(cmd);
        self
    }

    // -----------------------------------------------------------------------
    // Dispatch
    // -----------------------------------------------------------------------

    /// Main entry point called by the bot for every message / command.
    ///
    /// Resolution order:
    /// 1. Split on whitespace → first token → look up in registry.
    /// 2. If found, execute the command.
    /// 3. If not found and Ollama is available, ask the LLM.
    /// 4. Otherwise return a canned "unknown" response.
    pub async fn dispatch(&self, ctx: &Context, input: &str) -> Result<Response> {
        let input = input.trim();
        if input.is_empty() {
            return Ok(Response::Empty);
        }

        let parts: Vec<&str> = input.split_whitespace().collect();
        let token = parts[0].trim_start_matches('/');
        let args = &parts[1..];

        debug!("Agent dispatch: token={} args_len={}", token, args.len());

        // 1. Named command?
        if let Some(cmd) = self.registry.get(token) {
            return cmd.execute(ctx, args, self.state.clone()).await;
        }

        // 2. Starts with '/' but not registered → unknown command
        if input.starts_with('/') {
            return Ok(Response::text(format!(
                "Unknown command: /{}. Use /help to see available commands.",
                token
            )));
        }

        // 3. Free-text → Ollama if available
        self.ollama_fallback(ctx, input).await
    }

    /// Call Ollama with session context prepended.
    async fn ollama_fallback(&self, ctx: &Context, query: &str) -> Result<Response> {
        match &self.ollama {
            Some(client) => {
                // Build a short conversation history from session context.
                let session = self.state.get_or_create_session(ctx.chat_id).await;
                let context_window: Vec<String> = session
                    .context
                    .iter()
                    .rev()
                    .take(self.config.max_context_entries)
                    .rev()
                    .cloned()
                    .collect();

                let prompt = if context_window.is_empty() {
                    query.to_owned()
                } else {
                    format!(
                        "Context (recent messages):\n{}\n\nUser: {}",
                        context_window.join("\n"),
                        query
                    )
                };

                debug!("Sending to Ollama: {} chars", prompt.len());
                let response = client.complete(&prompt).await?;

                // Update session context
                self.state
                    .update_session(ctx.chat_id, |s| {
                        s.push_context(
                            format!("User: {}", query),
                            self.config.max_context_entries * 2,
                        );
                        s.push_context(
                            format!("Bot: {}", &response),
                            self.config.max_context_entries * 2,
                        );
                    })
                    .await;

                Ok(Response::Text(response))
            }
            None => {
                if self.config.silent_unknown {
                    Ok(Response::Empty)
                } else {
                    Ok(Response::text(
                        "I'm not sure how to help with that. \
                         Use /help to see available commands.",
                    ))
                }
            }
        }
    }

    // -----------------------------------------------------------------------
    // Accessors
    // -----------------------------------------------------------------------

    pub fn has_ollama(&self) -> bool {
        self.ollama.is_some()
    }

    pub fn registry(&self) -> &CommandRegistry {
        &self.registry
    }

    pub fn state(&self) -> Arc<AppState> {
        self.state.clone()
    }
}

// ---------------------------------------------------------------------------
// Built-in AgentCommand implementations
// ---------------------------------------------------------------------------

/// A pass-through that delegates to any `Fn` closure — useful for wiring up
/// thin adapters from `Command` implementations.
pub struct FnAgentCommand {
    name: String,
    description: String,
    aliases: Vec<String>,
    #[allow(clippy::type_complexity)]
    handler: Arc<
        dyn Fn(
                Context,
                Vec<String>,
                Arc<AppState>,
            ) -> std::pin::Pin<Box<dyn std::future::Future<Output = Result<Response>> + Send>>
            + Send
            + Sync,
    >,
}

impl FnAgentCommand {
    pub fn new<F, Fut>(
        name: impl Into<String>,
        description: impl Into<String>,
        handler: F,
    ) -> Arc<Self>
    where
        F: Fn(Context, Vec<String>, Arc<AppState>) -> Fut + Send + Sync + 'static,
        Fut: std::future::Future<Output = Result<Response>> + Send + 'static,
    {
        Arc::new(Self {
            name: name.into(),
            description: description.into(),
            aliases: Vec::new(),
            handler: Arc::new(move |ctx, args, state| Box::pin(handler(ctx, args, state))),
        })
    }
}

#[async_trait]
impl AgentCommand for FnAgentCommand {
    fn name(&self) -> &str {
        &self.name
    }

    fn description(&self) -> &str {
        &self.description
    }

    fn aliases(&self) -> &[&str] {
        // SAFETY: self.aliases is owned String vec; we return &str refs into it.
        // Since FnAgentCommand is not Copy this is fine as long as self lives.
        // But &[&str] requires a static or self-tied lifetime.
        // For simplicity we return empty here; aliases on this type are set via constructor.
        &[]
    }

    async fn execute(
        &self,
        ctx: &Context,
        args: &[&str],
        state: Arc<AppState>,
    ) -> Result<Response> {
        let args_owned: Vec<String> = args.iter().map(|s| s.to_string()).collect();
        (self.handler)(ctx.clone(), args_owned, state).await
    }
}

// ---------------------------------------------------------------------------
// Tests
// ---------------------------------------------------------------------------

#[cfg(test)]
mod tests {
    use super::*;
    use crate::state::BotConfig;
    use crate::types::Message;

    fn make_state() -> Arc<AppState> {
        AppState::new(BotConfig { token: "test".into(), ..Default::default() })
    }

    fn ctx(text: &str) -> Context {
        Context::new(Message::new(1, 1, text))
    }

    // Minimal AgentCommand for testing
    struct PingCmd;

    #[async_trait]
    impl AgentCommand for PingCmd {
        fn name(&self) -> &str { "ping" }
        fn description(&self) -> &str { "pong" }
        async fn execute(
            &self,
            _ctx: &Context,
            _args: &[&str],
            _state: Arc<AppState>,
        ) -> Result<Response> {
            Ok(Response::text("pong"))
        }
    }

    #[tokio::test]
    async fn dispatch_registered_command() {
        let state = make_state();
        let agent = Agent::new(state).register(Arc::new(PingCmd));
        let resp = agent.dispatch(&ctx("/ping"), "/ping").await.unwrap();
        assert_eq!(resp.as_str(), "pong");
    }

    #[tokio::test]
    async fn unknown_slash_command() {
        let state = make_state();
        let agent = Agent::new(state);
        let resp = agent.dispatch(&ctx("/foo"), "/foo").await.unwrap();
        assert!(resp.as_str().contains("Unknown command"));
    }

    #[tokio::test]
    async fn free_text_no_ollama() {
        let state = make_state();
        let agent = Agent::new(state);
        let resp = agent.dispatch(&ctx("hello"), "hello world").await.unwrap();
        assert!(!resp.as_str().is_empty());
    }

    #[tokio::test]
    async fn empty_input_returns_empty() {
        let state = make_state();
        let agent = Agent::new(state);
        let resp = agent.dispatch(&ctx(""), "").await.unwrap();
        assert!(resp.is_empty());
    }

    #[tokio::test]
    async fn registry_len() {
        let mut reg = CommandRegistry::new();
        reg.register(Arc::new(PingCmd));
        assert!(!reg.is_empty());
        assert!(reg.get("ping").is_some());
    }
}
